In [91]:
import psycopg2
dbname = 'kunal'
user = 'kunal'
password = 'kunalsoni'
host = 'localhost'
port = '5432' 
conn_string = f"dbname={dbname} user={user} password={password} host={host} port={port}"
try:
    connection = psycopg2.connect(conn_string)
    print("Connected to the database!")
except Exception as e:
    print(f"Unable to connect to the database. Error: {e}")
cursor = connection.cursor()

Connected to the database!


In [ ]:
##Top 5 Movie Titles based on Duration:

In [92]:
# Execute SQL query to get top 5 movies by duration
query_duration = "SELECT title, minutes FROM movies m where minutes is not null ORDER BY minutes DESC LIMIT 5;"
cursor.execute(query_duration)

# Fetch and print the results
results_duration = cursor.fetchall()
print("Top 5 Movies by Duration:") 
for row in results_duration:
    print(row)

Top 5 Movies by Duration:
('Gangs of Wasseypur', 320)
('Gone with the Wind', 238)
('Once Upon a Time in America', 229)
('Lawrence of Arabia', 216)
('Jodhaa Akbar', 213)


In [51]:
query_year = "SELECT title, year FROM movies where year is not null ORDER BY year DESC LIMIT 5;"
cursor.execute(query_year)

# Fetch and print the results
results_year = cursor.fetchall()
print("\nTop 5 Movies by Year of Release:")
for row in results_year:
    print(row)


Top 5 Movies by Year of Release:
('Bridge of Spies', 2015)
('Irrational Man', 2015)
('Ted 2', 2015)
('McFarland, USA', 2015)
('Cobain: Montage of Heck', 2015)


In [52]:
# Execute SQL query to get top 5 movies by average rating
query_avg_rating = """
    SELECT m.title, round(AVG(r.rating),2) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    GROUP BY m.title
    HAVING COUNT(r.rating) >= 5
    ORDER BY avg_rating DESC
    LIMIT 5;
"""
cursor.execute(query_avg_rating)

# Fetch and print the results
results_avg_rating = cursor.fetchall()
print("\nTop 5 Movies by Average Rating (min 5 ratings):")
for row in results_avg_rating:
    print(row)



Top 5 Movies by Average Rating (min 5 ratings):
('The Godfather: Part II', Decimal('9.80'))
('L.A. Confidential', Decimal('9.80'))
('The Shawshank Redemption', Decimal('9.65'))
('The Lord of the Rings: The Return of the King', Decimal('9.60'))
('Star Wars: Episode V - The Empire Strikes Back', Decimal('9.57'))


In [54]:
# Execute SQL query to get top 5 movies by number of ratings
query_num_ratings = """
    SELECT m.title, COUNT(r.rating) as num_ratings
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    GROUP BY m.title
    ORDER BY num_ratings DESC
    LIMIT 5;
"""
cursor.execute(query_num_ratings)

# Fetch and print the results
results_num_ratings = cursor.fetchall()
print("\nTop 5 Movies by Number of Ratings:")
for row in results_num_ratings:
    print(row)


Top 5 Movies by Number of Ratings:
('The Wolf of Wall Street', 69)
('Gravity', 64)
('Man of Steel', 60)
('Now You See Me', 59)
('Interstellar', 58)


In [55]:
# Execute SQL query to get the count of unique rater IDs
query_unique_raters = "SELECT COUNT(DISTINCT rater_id) FROM ratings r;"
cursor.execute(query_unique_raters)

# Fetch the result
unique_raters_count = cursor.fetchone()[0]

# Print the result
print(f"Number of Unique Raters: {unique_raters_count}")


Number of Unique Raters: 1048


In [56]:
# Execute SQL query to get top 5 raters by most movies rated
query_most_movies_rated = """
    SELECT rater_id, COUNT(movie_id) as movies_rated
    FROM ratings
    GROUP BY rater_id
    ORDER BY movies_rated DESC
    LIMIT 5;
"""
cursor.execute(query_most_movies_rated)

# Fetch and print the results
results_most_movies_rated = cursor.fetchall()
print("Top 5 Rater IDs by Most Movies Rated:")
for row in results_most_movies_rated:
    print(row)

Top 5 Rater IDs by Most Movies Rated:
(735, 314)
(769, 209)
(512, 188)
(800, 186)
(82, 159)


In [61]:
# Execute SQL query to get top 5 raters by highest average rating (min 5 ratings)
query_highest_avg_rating = """
    SELECT rater_id, round(AVG(rating),2) as avg_rating
    FROM ratings
    GROUP BY rater_id
    HAVING COUNT(rating) >= 5
    ORDER BY avg_rating DESC
    LIMIT 5;
"""
cursor.execute(query_highest_avg_rating)

# Fetch and print the results
results_highest_avg_rating = cursor.fetchall()
print("\nTop 5 Rater IDs by Highest Average Rating (min 5 ratings):")
for row in results_highest_avg_rating:
    print(row)



Top 5 Rater IDs by Highest Average Rating (min 5 ratings):
(256, Decimal('10.00'))
(231, Decimal('10.00'))
(662, Decimal('10.00'))
(206, Decimal('10.00'))
(146, Decimal('10.00'))


In [94]:
# Execute SQL query to get the top-rated movies by Michael Bay
query_top_rated_director = """
    SELECT m.title, round(AVG(r.rating),2) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE m.director = 'Michael Bay'
    GROUP BY m.title
    ORDER BY avg_rating DESC
    LIMIT 1;
"""
cursor.execute(query_top_rated_director)

# Fetch and print the result
result_top_rated_director = cursor.fetchone()
print("Top Rated Movie by Michael Bay:")
print(result_top_rated_director)

# Execute SQL query to get the top-rated comedy movie in the year 2013
query_top_rated_comedy = """
    SELECT m.title, round(AVG(r.rating),2) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE lower(m.genre) = 'comedy' 
    GROUP BY m.title
    ORDER BY avg_rating DESC
    LIMIT 1;
"""
cursor.execute(query_top_rated_comedy)

# Fetch and print the result
result_top_rated_comedy = cursor.fetchone()
print("\nTop Rated Comedy:")
print(result_top_rated_comedy)

# Execute SQL query to get the top-rated comedy movie in the year 2013
query_top_rated_2013 = """
    SELECT m.title, round(AVG(r.rating),2) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE  m.year = 2013
    GROUP BY m.title
    ORDER BY avg_rating DESC
    LIMIT 1;
"""
cursor.execute(query_top_rated_2013)

# Fetch and print the result
result_top_rated_2013 = cursor.fetchone()
print("\nTop Rated Movie in 2013:")
print(result_top_rated_2013)

# Execute SQL query to get the top-rated movie in India
query_top_rated_in_india = """
    SELECT m.title, round(AVG(r.rating),2) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE lower(m.country) = 'india'
    GROUP BY m.title
    HAVING COUNT(r.rating) >= 5
    ORDER BY avg_rating DESC
    LIMIT 1;
"""
cursor.execute(query_top_rated_in_india)

# Fetch and print the result
result_top_rated_in_india = cursor.fetchone()
print("\nTop Rated Movie in India:")
print(result_top_rated_in_india)



Top Rated Movie by Michael Bay:
('Armageddon', Decimal('10.00'))

Top Rated Comedy:
('A Very Harold & Kumar 3D Christmas', Decimal('10.00'))

Top Rated Movie in 2013:
('Hawking', Decimal('10.00'))

Top Rated Movie in India:
('PK', Decimal('9.30'))


In [90]:
# Execute SQL query to get the favorite movie genre for rater ID 1040
query_favorite_genre = """
    SELECT m.genre, COUNT(*) as genre_count
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE r.rater_id = 1040
    GROUP BY m.genre
    ORDER BY genre_count DESC
    LIMIT 1;
"""
cursor.execute(query_favorite_genre)

# Fetch and print the result
result_favorite_genre = cursor.fetchone()
if result_favorite_genre:
    favorite_genre, genre_count = result_favorite_genre
    print(f"Favorite Movie Genre for Rater ID 1040: {favorite_genre} (Rated {genre_count} times)")
else:
    print("Rater ID 1040 has not rated any movies.")

Favorite Movie Genre for Rater ID 1040: Action, Adventure, Sci-Fi (Rated 11 times)


In [85]:
# Execute SQL query to get the highest average rating for a movie genre by Rater ID 1040
query_highest_avg_rating = """
    SELECT m.genre, AVG(r.rating) as avg_rating
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE r.rater_id = 1040
    GROUP BY m.genre
    HAVING COUNT(r.rating) >= 5
    ORDER BY avg_rating DESC
    LIMIT 1;
"""
cursor.execute(query_highest_avg_rating)

# Fetch and print the result
result_highest_avg_rating = cursor.fetchone()
if result_highest_avg_rating:
    genre, highest_avg_rating = result_highest_avg_rating
    print(f"Highest Average Rating for a Movie Genre by Rater ID 1040:")
    print(f"Genre: {genre}, Average Rating: {highest_avg_rating:.2f}")
else:
    print("Rater ID 1040 has not rated enough movies in any genre.")


Highest Average Rating for a Movie Genre by Rater ID 1040:
Genre: Action, Adventure, Sci-Fi, Average Rating: 8.45


In [86]:
# Execute SQL query to get the year with the second-highest number of action movies
query_second_highest_action_movies = """
    SELECT m.year, COUNT(*) as num_action_movies
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE m.genre = 'Action' AND m.country = 'USA' AND m.minutes < 120
    GROUP BY m.year
    HAVING AVG(r.rating) >= 6.5
    ORDER BY num_action_movies DESC
    OFFSET 1
    LIMIT 1;
"""
cursor.execute(query_second_highest_action_movies)

# Fetch and print the result
result_second_highest_action_movies = cursor.fetchone()
if result_second_highest_action_movies:
    year, num_action_movies = result_second_highest_action_movies
    print(f"Year with Second-Highest Number of Action Movies:")
    print(f"Year: {year}, Number of Action Movies: {num_action_movies}")
else:
    print("No data found for the specified criteria.")

No data found for the specified criteria.


In [87]:
# Execute SQL query to count movies with high ratings
query_high_rated_movies_count = """
    SELECT COUNT(DISTINCT m.id)
    FROM movies m
    JOIN ratings r ON m.id = r.movie_id
    WHERE r.rating >= 7
    GROUP BY m.id
    HAVING COUNT(r.rating) >= 5;
"""
cursor.execute(query_high_rated_movies_count)

# Fetch and print the result
result_high_rated_movies_count = cursor.fetchone()
if result_high_rated_movies_count:
    high_rated_movies_count = result_high_rated_movies_count[0]
    print(f"Number of Movies with High Ratings (at least five reviews with a rating of 7 or higher): {high_rated_movies_count}")
else:
    print("No movies found that meet the specified criteria.")


Number of Movies with High Ratings (at least five reviews with a rating of 7 or higher): 1


Top 5 Movies by Duration:
'Gangs of Wasseypur' (320 minutes)
'Gone with the Wind' (238 minutes)
'Once Upon a Time in America' (229 minutes)
'Lawrence of Arabia' (216 minutes)
'Jodhaa Akbar' (213 minutes)

Top 5 Movies by Year of Release:
'Bridge of Spies' (2015)
'Irrational Man' (2015)
'Ted 2' (2015)
'McFarland, USA' (2015)
'Cobain: Montage of Heck' (2015)

Top 5 Movies by Average Rating (min 5 ratings):
'The Godfather: Part II' (9.80)
'L.A. Confidential' (9.80)
'The Shawshank Redemption' (9.65)
'The Lord of the Rings: The Return of the King' (9.60)
'Star Wars: Episode V - The Empire Strikes Back' (9.57)

Top 5 Movies by Number of Ratings:
'The Wolf of Wall Street' (69)
'Gravity' (64)
'Man of Steel' (60)
'Now You See Me' (59)
'Interstellar' (58)

Number of Unique Raters:
1048

Top 5 Rater IDs by Most Movies Rated:
(735, 314)
(769, 209)
(512, 188)
(800, 186)
(82, 159)

Top 5 Rater IDs by Highest Average Rating (min 5 ratings):
(256, 10.00)
(231, 10.00)
(662, 10.00)
(206, 10.00)
(146, 10.00)

Top Rated Movie by Michael Bay:
'Armageddon' (10.00)

Top Rated Comedy:
'A Very Harold & Kumar 3D Christmas' (10.00)

Top Rated Movie in 2013:
'Hawking' (10.00)

Top Rated Movie in India:
'PK' (9.30)

Favorite Movie Genre for Rater ID 1040:
Action, Adventure, Sci-Fi (Rated 11 times)

Highest Average Rating for a Movie Genre by Rater ID 1040:
Genre: Action, Adventure, Sci-Fi, Average Rating: 8.45
                      
Number of Movies with High Ratings (at least five reviews with a rating of 7 or higher):
1
                      
Insights:

The movie 'Gangs of Wasseypur' is the longest among the top 5 movies by duration, lasting 320 minutes.

The year 2015 had a significant number of top movies, including 'Bridge of Spies' and 'Irrational Man.'

Movies like 'The Godfather: Part II' and 'L.A. Confidential' stand out with high average ratings.

Raters with IDs 735 and 769 have rated the most movies, while raters like 256 and 231 have consistently given perfect ratings.

Michael Bay's 'Armageddon' and the comedy 'A Very Harold & Kumar 3D Christmas' received perfect ratings.

The movie 'Hawking' in 2013 and 'PK' in India were highly rated.

Rater ID 1040 has a preference for the genre of Action, Adventure, Sci-Fi, rating it the most.

The highest average rating for a movie genre by Rater ID 1040 is for Action, Adventure, Sci-Fi, with an average rating of 8.45.

Only one movie has received high ratings (at least five reviews with a rating of 7 or higher).